### DISCUSSION TITLE
*try catch* problem: `$err.raw` stops execution

### DISCUSSION THREAD

**Discussion by askeladd123** [nushell/nushell#14080] [Q&A] [answered]
**Question:** how can I print an error like normal but continue execution?

When I run: 
```
try { 1 / 0} catch { |err| print $err}; print hello
```
... the error record is printed, followed by "hello".

But when I run:
```
try { 1 / 0} catch { |err| print $err.raw}; print hello
```
... only the raw error is printed, and execution stops. `print hello` is never reached. 

The reason I want to print `$err.raw` is to print the same error message as if the program failed. I want to keep the formatting, the colors and the full context, none of which can be accomplished with `$e.msg` or `$e.debug`.

## from documentation
From the *guide* on the *Error* data type:
> ... Any error condition results in termination of the current command/expression and will also terminate any assignment expression or custom command. 

I realize this, but how can I print the error then, if the error cannot be touched without crashing the program?

## use case
I am writing a nu script for testing a cli. If the test fails, I want to print the errors as usual, record the status of the test, and continue to the next one. Currently the test will stop as soon as one of them fail, which is not ideal. 

This problem aside, writing tests for cli's feels nice in Nushell. Keep it up :wink:

**Comment by ghost** [comment]
Im not entirely sure on the why, but from my observation any access of `$err.raw` re-throws the error.  
i was unable to find a way to use the integrated error-renderer, but `$err.msg` + `$err.debug` is often good enough.  
You can try to parse `$debug` and try to render it yourself, but well the parsing isnt reliable and the rendering isnt easy either.

many (not all) errors can be parsed via the following:
```nu
try {
  non_existent_command
} catch {|err|
  print ($'($err.debug | str substring ($err.debug | str index-of "{")..(($err.debug | str index-of "span:") - 3))}' | from nuon)
}
```

**Reply by fdncred** [reply]
I was able to do this. It's as close as I can get. I'll try to put up a PR later.
![image](https://github.com/user-attachments/assets/3e7a841e-71fa-4832-bd0d-06c8cc4c638c)

**Reply by askeladd123** [reply]
Ty for quick respnse! So this new `shell_error` will print the same as `raw_err`, without stopping execution?

**Reply by fdncred** [reply, marked as answer]
Here's the PR https://github.com/nushell/nushell/pull/14082. `raw` is the error itself and will act as such if you try to touch it. I changed shell_error to rendered and added json to be able to easily parse everything out with the ability to use `error make` manually if you want.

**Comment by ghost** [comment]
I know its now possible in 0.99, but i just noticed an option to print it in >=0.98 (if you want to keep something backward-compatible):

```
try { non_existent_command } catch {|e| 0 | tee {|| $e.raw} | null }
```

this works, since `tee` outputs errors to stderr without crashing since 0.98 ([docs](https://www.nushell.sh/blog/2024-09-17-nushell_0_98_0.html#tee-toc))

**Reply by askeladd123** [reply]
This worked! Only problem I can think of is that `tee` runs on another thread, so `tee { sleep 1sec; print foo}; print bar` will print `bar foo`. 

---

> I know its now possible in 0.99 ...

Also #14082 is not merged yet right?

**Reply by ghost** [reply]
> Only problem I can think of is that tee runs on another thread, so tee { sleep 1sec; print foo}; print bar will print bar foo.

you could add `; sleep 0.25sec` after `tee` - not a guarantee, but pretty reliable. yes it will slow it down, but only in a error-case

> Also #14082 is not merged yet right?

right, sorry - i saw `try` in the TOC of the changelog and somehow associated it with this in the back of my head
